[quant][graphmode][fix] insert_prepack_unpack pattern

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #27119 [WIP][quant][graphmode] Prepack folding for conv2d
* #27118 [quant][graphmode] Add insert_prepack_unpack for conv2d
* #27103 [quant][graphmode] Enable tests
* **#27102 [quant][graphmode][fix] insert_prepack_unpack pattern**
* #27093 [quant][graphmode] API - add more passes to graph mode

Summary:
We need to prepack the quantized weight rather then original weight

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D17678264](https://our.internmc.facebook.com/intern/diff/D17678264)

jerryzh168 committed:
[quant][graphmode][fix] insert_prepack_unpack pattern

Summary:
We need to prepack the quantized weight rather then original weight

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] insert_prepack_unpack pattern"


Summary:
We need to prepack the quantized weight rather then original weight

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -869,26 +869,35 @@ graph(%self, %scale, %zero_point, %dtype):
+                   const std::unordered_map<std::string, Value*>& vmap) {
+     const auto& match_vmap = match.values_map;
+     auto linear_node = match_vmap.at(vmap.at("linear"))->node();
+     if (linear_node->kind() == prim::Constant && linear_node->s(attr::name) == "linear") {
```
Can we actually check the type rather than relying on a decorative attribute in the `prim::Constant` node?

jerryzh168 replied:
Sure

jerryzh168 committed:
Update on "[quant][graphmode][fix] insert_prepack_unpack pattern"


Summary:
We need to prepack the quantized weight rather then original weight

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -869,26 +869,37 @@ graph(%self, %scale, %zero_point, %dtype):
+     auto linear_node = match_vmap.at(vmap.at("linear"))->node();
+     auto func = linear_node->output()->type()->expect<FunctionType>()->function();
+     auto func_name = getFuncName(func->qualname());
+     if (linear_node->kind() == prim::Constant && func_name == "linear") {
```
I think we can remove `linear_node->kind() == prim::Constant` check, checking the type should be sufficient. In theory, this value can be passed as an argument to the function, in which case the node won't be `Constant`, but the function will still be `linear` so the transformation would be valid.

jerryzh168 replied:
I'll remove Constant check for another callsite as well

ZolotukhinM reviewed (approved):
Thanks, looks good now!

jerryzh168 committed:
Update on "[quant][graphmode][fix] insert_prepack_unpack pattern"


Summary:
We need to prepack the quantized weight rather then original weight

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D17678264](https://our.internmc.facebook.com/intern/diff/D17678264)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fix] insert_prepack_unpack pattern"


Summary:
We need to prepack the quantized weight rather then original weight

Test Plan:
.

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D17678264](https://our.internmc.facebook.com/intern/diff/D17678264)

[ghstack-poisoned]
